ui/app/contexts/metametrics.js

```diff
@@ -1,24 +1,101 @@
-} from '../../../selectors'
-
-import { getEnvironmentType } from '../../../../../app/scripts/lib/util'
+} from '../selectors/confirm-transaction'
+import { getEnvironmentType } from '../../../app/scripts/lib/util'
 import {
   sendMetaMetricsEvent,
   sendCountIsTrackable,
-} from '../../utils/metametrics.util'
+} from '../helpers/utils/metametrics.util'
+
+
+export const MetaMetricsContext = createContext(() => {})
+
+export function MetaMetricsProvider ({ children }) {
+  const txData = useSelector(txDataSelector) || {}
+  const network = useSelector(getCurrentNetworkId)
+  const environmentType = getEnvironmentType()
+  const activeCurrency = useSelector(getSelectedAsset)
+  const accountType = useSelector(getAccountType)
+  const confirmTransactionOrigin = txData.origin
+  const metaMetricsId = useSelector((state) => state.metamask.metaMetricsId)
+  const participateInMetaMetrics = useSelector((state) => state.metamask.participateInMetaMetrics)
+  const metaMetricsSendCount = useSelector((state) => state.metamask.metaMetricsSendCount)
+  const numberOfTokens = useSelector(getNumberOfTokens)
+  const numberOfAccounts = useSelector(getNumberOfAccounts)
+  const history = useHistory()
+  const [state, setState] = useState(() => ({
+    currentPath: window.location.href,
+    previousPath: '',
+  }))
+
+  const { previousPath, currentPath } = state
+
+  useEffect(() => {
+    const unlisten = history.listen(() => setState((prevState) => ({
+      currentPath: window.location.href,
+      previousPath: prevState.currentPath,
+    })))
+    // remove this listener if the component is no longer mounted
+    return unlisten
+  }, [history])
+
+  const metricsEvent = useCallback((config = {}, overrides = {}) => {
+    const { eventOpts = {} } = config
+    const { name = '' } = eventOpts
+    const { pathname: overRidePathName = '' } = overrides
+    const isSendFlow = Boolean(name.match(/^send|^confirm/) || overRidePathName.match(/send|confirm/))
+
+    if (participateInMetaMetrics || config.isOptIn) {
+      return sendMetaMetricsEvent({
+        network,
+        environmentType,
+        activeCurrency,
+        accountType,
+        confirmTransactionOrigin,
+        metaMetricsId,
+        numberOfTokens,
+        numberOfAccounts,
+        ...config,
+        previousPath,
+        currentPath,
+        excludeMetaMetricsId: isSendFlow && !sendCountIsTrackable(metaMetricsSendCount + 1),
+        ...overrides,
+      })
+    }
+  }, [
+    network,
+    environmentType,
+    activeCurrency,
+    accountType,
+    confirmTransactionOrigin,
+    metaMetricsId,
+    numberOfTokens,
+    numberOfAccounts,
+    currentPath,
+  ])
+
+  return (
+    <MetaMetricsContext.Provider value={metricsEvent}>
+      {children}
+    </MetaMetricsContext.Provider>
+  )
+}
+
+MetaMetricsProvider.propTypes = { children: PropTypes.node.isRequired }
 
-class MetaMetricsProvider extends Component {
+class LegacyMetaMetricsProvider extends Component {
```

Comment:
Could you update this legacy provider to re-use the new hooks-based provider? I did a similar thing in `i18n.js` - this class can consume the new context API and expose the legacy one.

Reply:
@Gudahtt -- mind blown. That would, infact, make this a pretty important PR to get full QA on to make sure analytics are not going to be impacted right because it would just proxy the context to the new API and if I inadvertently f%$#ed it up we got problems?

Reply:
Yeah I guess so! It's not particularly easy to QA, but we could certainly review it more carefully. Maybe we could compare the network logs before and after to verify they're equivalent.